### PR TITLE
restructure the setup and teardown functions for network.py scripts (Bugfix)

### PR DESCRIPTION
**Pull request by stanley31huang**
## Description
There are some improvement in this PR as following:
1. restructure the setup and teardown functions with context manager library. 
2. not bring up the network interfaces if it was down before testing.
3. not shutdown the conduit network interface (master network interface) while testing with an user network interface (slave network interface) (for Linux DSA network)

## Resolved issues
https://github.com/canonical/checkbox/issues/233
https://github.com/canonical/checkbox/issues/401

## Documentation
Reference:
- https://dri.freedesktop.org/docs/drm/networking/dsa/dsa.html
- https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-class-net

### Tested on system with multiple physical network
- all passed results: 
https://certification.canonical.com/hardware/202405-34042/submission/391153/
- eno3 network interface failed as expected (no ethernet cable connected), it's to simulate the test failed and make sure this scripts could restore the network configuration back. 
https://certification.canonical.com/hardware/202405-34042/submission/381308/

### Tested on system with DSA network
- the lan2 network interface failed as expected, it's to simulate the iperf server is not stable and make sure this scripts could restore the network configuration back.
https://certification.canonical.com/hardware/202405-34042/submission/381472/

### CODE REVIEW

File: providers/base/bin/network.py [hunk 1]
```diff
@@ -753,24 +767,208 @@ def make_target_list(iface, test_targets, log_warnings):
...
+        cur_network_info = get_network_ifaces()
+        recover_success = restore_network_ifaces(
+            cur_network_info, network_info, recover_timeout
+        )
+
+        # Restore routing table to original state
+        logging.debug("Restore routing table")
+        try:
+            # Harmless "RTNETLINK answers: File exists" messages on stderr
+            with open(os.devnull, "wb") as DEVNULL:
```
Comment (Hook25) [inline providers/base/bin/network.py:958]: Please use: https://docs.python.org/3/library/subprocess.html#subprocess.DEVNULL

File: providers/base/bin/network.py [hunk 2]
```diff
@@ -753,24 +767,208 @@ def make_target_list(iface, test_targets, log_warnings):
...
+
+    finally:
+        cur_network_info = get_network_ifaces()
+        recover_success = restore_network_ifaces(
+            cur_network_info, network_info, recover_timeout
+        )
+
+        # Restore routing table to original state
+        logging.debug("Restore routing table")
+        try:
```
Comment (Hook25) [inline providers/base/bin/network.py:956]: Like the comment! ty, I would use `with suppress(CalledProcessError)` or alternatively... `run` istead of `check_call` there is no point to using `check_call` if you ignore the check!

**Review by Hook25 — CHANGES_REQUESTED** [review]
For me it is very hard to understand if this is ok actually, I have a few nitpicks about the code but nothing major. The submissions you provide actually have 2 failed tests, so I'm not sure what is going on there. Could this just randomly fail on some machines? Did you test on just two? Can you please test this on more? This is not ran during sru right?

File: providers/base/bin/network.py (hunk 2 above)
Comment (stanley31huang) [reply providers/base/bin/network.py]: thanks for the suggestion, I have applied it into the latest commit

File: providers/base/bin/network.py (hunk 1 above)
Comment (stanley31huang) [reply providers/base/bin/network.py]: fixed

**Comment by stanley31huang** [comment]
> For me it is very hard to understand if this is ok actually, I have a few nitpicks about the code but nothing major. The submissions you provide actually have 2 failed tests, so I'm not sure what is going on there. 

> Could this just randomly fail on some machines? Did you test on just two? Can you please test this on more? 

I am sorry for the confusion on those failed test cases, it's trying to simulate some test failed scenario and make sure the network configuration could be restore. I have added more description for those submissions with failed tests, and attached a new submission without failed tests. 

> This is not ran during sru right?

I think so, it's not include in the SRU test plan as this is only include in the stress test plan.

> 
> (Oh and fantastic job with the context manager, I really like that!)

**Review by Hook25 — APPROVED** [review]

**Pull request merged**